### ISSUE TITLE
`test_strftime_y2k` fails on embedded Linux

### ISSUE THREAD

**Issue by picnixz** [python/cpython#123681]
# Bug report

### Bug description:

This is an issue to track the progress of fixing the JIT. Branches that trigger the embedded Linux tests fail on `test_strftime_y2k`, e.g. https://github.com/python/cpython/actions/runs/10655985897/job/29534232332?pr=123546. As a side-effect this causes users getting notification that their JIT workflow failed. 

Both the C and F specifiers are failing for this specific test:

<details>

```
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) ... 
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=1, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=1, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=48, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=48, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=70, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=70, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=999, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDate_Fast.test_strftime_y2k) (year=999, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) ... 
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=1, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=1, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=48, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=48, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=70, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=70, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=999, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTime_Fast.test_strftime_y2k) (year=999, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) ... 
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=1, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=1, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=48, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=48, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=70, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=70, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=999, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestDateTimeTZ_Fast.test_strftime_y2k) (year=999, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) ... 
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=1, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=1, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=48, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=48, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=70, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=70, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=99, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=99, specifier='C') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=999, specifier='F') ... FAIL
test_strftime_y2k (test.datetimetester.TestSubclassDateTime_Fast.test_strftime_y2k) (year=999, specifier='C') ... FAIL
```
</details>

A solution is to add this test to the ignored tests or to fix the detection algorithm.

cc @blhsing @serhiy-storchaka 

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Other

### Linked PRs
* gh-123861
* gh-128444
* gh-136022
* gh-137947

Related and proposed PR: gh-128106

**Comment by serhiy-storchaka** [comment]
It seems that only tests for the C implementation fail.

We can try to move the detection from the compile time to the run time, like for Python implementation. But it is a puzzle why it does not work here.

**Comment by picnixz** [comment]
How about checking it `configure.ac`?

**Comment by picnixz** [comment]
(Forget my comment that I deleted, I looked at the wrong commit).

**Comment by brandtbucher** [comment]
This particular test failure could be a bit weird, since this CI job involves cross-compiling Python and then running the test suite under emulation. Not sure if it's a "real" issue, or if we should just skip the test (like we do for [others](https://github.com/python/cpython/blob/main/Tools/jit/ignore-tests-emulated-linux.txt)).

**Comment by picnixz** [comment]
> since this CI job involves cross-compiling Python and then running the test suite under emulation

Does it mean that the configure tests might not be correct? If so, we can just outright ignore the test I guess?

**Comment by serhiy-storchaka** [comment]
If these configure tests are not correct, then other configure tests can be not correct too.

**Comment by picnixz** [comment]
Ha! I knew I should have looked more in detail the logs:

On **aarch64-unknown-linux-gnu** (failing tests), we have **two** different configure:

https://productionresultssa4.blob.core.windows.net/actions-results/75994383-b5b8-44c0-b3e3-e9c98d3735af/workflow-job-run-f1b62a84-9cf9-5fdd-2848-13ec49660028/logs/job/job-logs.txt?rsct=text%2Fplain&se=2024-09-08T18%3A24%3A08Z&sig=tTaLFKOgDrb57d8sfh%2FZQrNS6Gh1lrlnt1%2FnUqymMLc%3D&ske=2024-09-09T05%3A56%3A51Z&skoid=ca7593d4-ee42-46cd-af88-8b886a2f84eb&sks=b&skt=2024-09-08T17%3A56%3A51Z&sktid=398a6654-997b-47e9-b12b-9515b896b4de&skv=2024-05-04&sp=r&spr=https&sr=b&st=2024-09-08T18%3A14%3A03Z&sv=2024-05-04


> 2024-09-08T16:49:27.5943515Z checking whether year with century should be normalized for strftime... __**yes**__
2024-09-08T16:49:27.6286869Z checking whether C99-specific strftime specifiers are supported... __**yes**__

> 2024-09-08T16:51:33.7240722Z checking whether year with century should be normalized for strftime... __**yes**__
2024-09-08T16:51:33.7242911Z checking whether C99-specific strftime specifiers are supported... __**>> no <<**__

So it appears we have different ./configure. I don't know which one is actually being used at runtime in the emulated environment. Now, both '%F' and '%C' are C99-specific (https://en.cppreference.com/w/c/chrono/strftime) so the following

```py
if _time.strftime('%F', (1900, 1, 1, 0, 0, 0, 0, 1, 0)) == '1900-01-01':
    specifiers += 'FC'
```

is perhaps a too broad check. Since specifiers 'F' and 'C' are used, we actually expect to support `%F`, but this contradicts the fact that "C99-specific strftime specifiers are supported" (at least for the second one which seems the most recent and probably the one being used the hood). The configure check is:

```C
if (strftime(full_date, sizeof(full_date), "%F", &date) && !strcmp(full_date, "1900-01-01")) {
    return 0;
}
```

In other words, either `strftime(full_date, sizeof(full_date), "%F", &date)` fails or `full_date != 1900-01-01`. So maybe `%F` does not fail. We can probably debug the tests to check whether it `_time.strftime` that bugs, or if it is the configure test that bugs. I don't know if it's possible for `_time.strftime` to support `%F` (which does not directly call `strftime()`) but at the same time fail the `strftime(full_date, sizeof(full_date), "%F", &date) && !strcmp(full_date, "1900-01-01")` test.

---

I'm tired so what I wrote here might be wrong but I'll try to investigate tomorrow.

**Comment by blhsing** [comment]
> > 2024-09-08T16:49:27.5943515Z checking whether year with century should be normalized for strftime... **yes**
> > 2024-09-08T16:49:27.6286869Z checking whether C99-specific strftime specifiers are supported... **yes**
> 
> > 2024-09-08T16:51:33.7240722Z checking whether year with century should be normalized for strftime... **yes**
> > 2024-09-08T16:51:33.7242911Z checking whether C99-specific strftime specifiers are supported... **>> no <<**
Ahh how could I have not noticed the second configure! Was scratching my head over how detection of C99 support got a yes but the compiled code ended up not handling C99. It makes total sense now.

The problem is simply that autoconf does not run the test for the [`AC_RUN_IFELSE`](https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Runtime.html) macro at all when cross-compiling, but directly performs *action-if-cross-compiling* as specified, which I thought should be made `no` when it should really be `yes` because we should make sure to handle `%F` and `%C` ourselves when we are unsure if the target platform can handle them.

**Comment by picnixz** [comment]
> Ahh how could I have not noticed the second configure in the logs!

Well, isn't it trivial to spot it among 11985 lines 😄? Actually, I found the duplicate configure in the logs by a CTRL+F and... saw that I had two results.

**Comment by blhsing** [comment]
> > Ahh how could I have not noticed the second configure in the logs!
> 
> Well, isn't it trivial to spot it among 11985 lines 😄? Actually, I found the duplicate configure in the logs by a CTRL+F and... saw that I had two results.

I used Ctrl-F too, but the browser limits the search to only what's loaded in the memory, when the output of the second configure is hidden because it is loaded on demand by an AJAX call only if I scroll to it. I should've used the "Search Logs" text box in the page to perform the search for the keyword on the server side instead.

Thanks again!

**Comment by picnixz** [comment]
> I used Ctrl-F too, but the browser limits the search to only what's loaded in the memory, when the output of the second configure is hidden because it is loaded on demand by an AJAX call only if I scroll to it. I should've used the "Search Logs" text box in the page to perform the search for the keyword on the server side instead.

Actually, the trick is to "view raw logs". You'll get the entire logs without buffering (easier to read IMO).

**Comment by picnixz** [comment]
FTR: https://github.com/python/cpython/pull/124466 was merged as a temporary workaround so we need to keep the issue opened even though the job looks successful (it's just that we skip it).

**Comment by picnixz** [comment]
Closing this one in favor of gh-128104 which has a better description and an alternative PR.

**Comment by serhiy-storchaka** [comment]
Why not test this at runtime instead of compile time, like in the Python implementation?

**Comment by serhiy-storchaka** [comment]
```pycon
>>> time.strftime("%C", (99, 1, 1, 0, 0, 0, 0, 1, 0))
'0'
```
In C99 it should be 2 digits.

**Comment by encukou** [comment]
So, we need a test for `normalize_century` (ideally at run-time, to avoid cross-compilation issues), but in CPython we can assume `%C` and `%F` are *available*, right?

**Comment by serhiy-storchaka** [comment]
They are in C99. https://en.cppreference.com/w/c/chrono/strftime

It seems that the main reason of adding checks for them (see #122272) is that they are not documented in Python, and therefore their support is not guaranteed. If we guarantee their support, we should also document them and add tests. We can make this a new feature in 3.15 or even in 3.14, but I have doubts about 3.13.

**Comment by encukou** [comment]
As I see it, they're guaranteed in CPython (via the C99 build requirement), but not in Python the language.

**Comment by serhiy-storchaka** [comment]
This is an interesting point of view. Then we can remove the check in the C code (which is CPython specific), but should leave it in the Python code, which  can be reused in other implementation. From this point of view, #128444 and #136022 are equivalent.

**Comment by encukou** [comment]
I think #136022 removing `def _supports_c99` is OK.
The Python code still uses `_time.strftime`. If other implementations replace `_time`, they will need to update the Python code as well. There's a test so they don't forget.
